src/agents/agent.py

```diff
@@ -63,42 +66,53 @@ def __init__(self, base_model: str, search_engine: str):
         self.engine = search_engine
         self.tokenizer = tiktoken.get_encoding("cl100k_base")
 
-    def search_queries(self, queries: list, project_name: str, engine: str) -> dict:
+    async def open_page(self, project_name, pdf_download_url):
+        browser = await Browser().start()
+
+        await browser.go_to(pdf_download_url)
```

Comment:
How can this better handle the case if a timeout has occurred and there is no page to evaluate?

This happens for me:

devika-backend-engine-1  | Searching in Google...
devika-backend-engine-1  | 
devika-backend-engine-1  | Link ::  https://gml.noaa.gov/grad/solcalc/azel.html 
devika-backend-engine-1  | 
devika-backend-engine-1  | TimeoutError: Timeout 20000ms exceeded. when trying to navigate to https://gml.noaa.gov/grad/solcalc/azel.html
devika-backend-engine-1  | Exception in thread Thread-6 (<lambda>):
devika-backend-engine-1  | Traceback (most recent call last):
devika-backend-engine-1  |   File "/usr/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
devika-backend-engine-1  |     self.run()
devika-backend-engine-1  |   File "/usr/lib/python3.11/threading.py", line 975, in run
devika-backend-engine-1  |     self._target(*self._args, **self._kwargs)
devika-backend-engine-1  |   File "/home/nonroot/devika/devika.py", line 96, in <lambda>
devika-backend-engine-1  |     thread = Thread(target=lambda: agent.execute(message, project_name))
devika-backend-engine-1  |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
devika-backend-engine-1  |   File "/home/nonroot/devika/src/agents/agent.py", line 348, in execute
devika-backend-engine-1  |     search_results = self.search_queries(queries, project_name)
devika-backend-engine-1  |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
devika-backend-engine-1  |   File "/home/nonroot/devika/src/agents/agent.py", line 111, in search_queries
devika-backend-engine-1  |     browser, raw, data = loop.run_until_complete(self.open_page(project_name, link))
devika-backend-engine-1  |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
devika-backend-engine-1  |   File "/usr/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
devika-backend-engine-1  |     return future.result()
devika-backend-engine-1  |            ^^^^^^^^^^^^^^^
devika-backend-engine-1  |   File "/home/nonroot/devika/src/agents/agent.py", line 73, in open_page
devika-backend-engine-1  |     _, raw = await browser.screenshot(project_name)
devika-backend-engine-1  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
devika-backend-engine-1  |   File "/home/nonroot/devika/src/browser/browser.py", line 42, in screenshot
devika-backend-engine-1  |     page_metadata = await self.page.evaluate("() => { return { url: document.location.href, title: document.title } }")
devika-backend-engine-1  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
devika-backend-engine-1  |   File "/home/nonroot/devika/.venv/lib/python3.11/site-packages/playwright/async_api/_generated.py", line 8119, in evaluate
devika-backend-engine-1  |     await self._impl_obj.evaluate(
devika-backend-engine-1  |   File "/home/nonroot/devika/.venv/lib/python3.11/site-packages/playwright/_impl/_page.py", line 416, in evaluate
devika-backend-engine-1  |     return await self._main_frame.evaluate(expression, arg)
devika-backend-engine-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
devika-backend-engine-1  |   File "/home/nonroot/devika/.venv/lib/python3.11/site-packages/playwright/_impl/_frame.py", line 278, in evaluate
devika-backend-engine-1  |     await self._channel.send(
devika-backend-engine-1  |   File "/home/nonroot/devika/.venv/lib/python3.11/site-packages/playwright/_impl/_connection.py", line 59, in send
devika-backend-engine-1  |     return await self._connection.wrap_api_call(
devika-backend-engine-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
devika-backend-engine-1  |   File "/home/nonroot/devika/.venv/lib/python3.11/site-packages/playwright/_impl/_connection.py", line 509, in wrap_api_call
devika-backend-engine-1  |     return await cb()
devika-backend-engine-1  |            ^^^^^^^^^^
devika-backend-engine-1  |   File "/home/nonroot/devika/.venv/lib/python3.11/site-packages/playwright/_impl/_connection.py", line 97, in inner_send
devika-backend-engine-1  |     result = next(iter(done)).result()
devika-backend-engine-1  |              ^^^^^^^^^^^^^^^^^^^^^^^^^
devika-backend-engine-1  | playwright._impl._errors.Error: Execution context was destroyed, most likely because of a navigation